tsl/src/chunk.c

```diff
@@ -134,32 +137,134 @@ chunk_set_foreign_server(Chunk *chunk, ForeignServer *new_server)
-	if (existing_server_id != foreign_table->serverid)
-		return;
+	/* Cannot switch to other data node if only one or none assigned */
+	if (list_length(chunk->data_nodes) < 2)
+		return false;
 
-	Assert(list_length(chunk->data_nodes) > 1);
+	/* Nothing to do if the chunk table already has the requested data node set */
+	if ((!available && data_node_id != foreign_table->serverid) ||
+		(available && data_node_id == foreign_table->serverid))
+		return true;
 
-	foreach (lc, chunk->data_nodes)
+	if (available)
 	{
-		new_server = lfirst(lc);
-		if (new_server->foreign_server_oid != existing_server_id)
-			break;
+		/* Switch to using the given data node, but only on chunks where the
+		 * given node is the "default" according to partitioning */
+		Cache *htcache = ts_hypertable_cache_pin();
+		const Hypertable *ht =
+			ts_hypertable_cache_get_entry(htcache, chunk->hypertable_relid, CACHE_FLAG_NONE);
+		const Dimension *dim = hyperspace_get_closed_dimension(ht->space, 0);
+
+		if (dim != NULL)
+		{
+			/* For space-partitioned tables, use the current partitioning
+			 * configuration in that dimension (dimension partition) as a
+			 * template for picking the query data node */
+			const DimensionSlice *slice =
+				ts_hypercube_get_slice_by_dimension_id(chunk->cube, dim->fd.id);
+			unsigned int i;
+
+			Assert(dim->dimension_partitions);
+
+			for (i = 0; i < dim->dimension_partitions->num_partitions; i++)
+			{
+				const DimensionPartition *dp = dim->dimension_partitions->partitions[i];
+
+				/* Match the chunk with the dimension partition. Count as a
+				 * match if the start of chunk is within the range of the
+				 * partition. This captures both the case when the chunk
+				 * aligns perfectly with the partition and when it is bigger
+				 * or smaller (due to a previous partitioning change). */
+				if (slice->fd.range_start >= dp->range_start &&
+					slice->fd.range_start <= dp->range_end)
+				{
+					ListCell *lc;
+
+					/* Use the data node for queries if it is the first
+					 * available data node in the partition's list (i.e., the
+					 * default choice) */
+					foreach (lc, dp->data_nodes)
+					{
+						const char *node_name = lfirst(lc);
+						server = GetForeignServerByName(node_name, false);
+
+						if (ts_data_node_is_available_by_server(server))
+						{
+							should_switch_data_node = (server->serverid == data_node_id);
+							break;
+						}
+					}
+				}
+			}
+		}
+		else
+		{
+			/* For hypertables without a space partition, use the data node
+			 * assignment logic to figure out whether to use the data node as
+			 * query data node. The "default" query data node is the first in
+			 * the list. The chunk assign logic only returns available data
+			 * nodes. */
+			List *datanodes = ts_hypertable_assign_chunk_data_nodes(ht, chunk->cube);
+			const char *node_name = linitial(datanodes);
+			server = GetForeignServerByName(node_name, false);
+
+			if (strcmp(server->servername, node_name) == 0)
```

Comment:
clearly this a bug @erimatnor? 

The node_name and server->servername will always match :-D

I believe we need

should_switch_data_node = (server->serverid == data_node_id);

Reply:
Yes, it should compare serverid with the function parameter data_node_id